engine/src/state_chain_observer/client.rs

```diff
@@ -491,23 +524,77 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
 
 					// Often `finalized_header` returns a significantly newer latest block than the
 					// stream returns so we move the stream forward to this block
-					{
-						let finalised_header_hash =
-							base_rpc_client.latest_finalized_block_hash().await?;
-						let finalised_header =
-							base_rpc_client.block_header(finalised_header_hash).await?;
-						if finalized_block_stream.cache().number < finalised_header.number {
-							let blocks_to_skip =
-								finalized_block_stream.cache().number + 1..=finalised_header.number;
-							for block_number in blocks_to_skip {
+					let finalized_hash = base_rpc_client.latest_finalized_block_hash().await?;
+					let finalized_header = base_rpc_client.block_header(finalized_hash).await?;
+
+					let base_rpc_client = base_rpc_client.clone();
+					let mut finalized_block_stream: Pin<
+						Box<
+							dyn TryCachedStream<
+									Ok = BlockInfo,
+									Error = anyhow::Error,
+									Item = Result<BlockInfo, anyhow::Error>,
+								> + Send,
+						>,
+					> = match start_from {
+						Some(start_from) if start_from < finalized_block_stream.cache().number => {
+							// save this to avoid fetching the same block again
+							let latest_block_from_cache = *finalized_block_stream.cache();
+
+							// we can make this range exclusive because we've saved the cached block
+							// above, which we chain later.
+							let earlier_blocks_to_fetch =
+								start_from..finalized_block_stream.cache().number;
+
+							let base_rpc_client = base_rpc_client.clone();
+							let mut start_from_headers = Box::pin(
+								futures::stream::iter(earlier_blocks_to_fetch)
+									.then(move |block_number| {
+										let base_rpc_client_c = base_rpc_client.clone();
+										async move {
+											let hash = base_rpc_client_c
+												.block_hash(block_number)
+												.await?
+												.ok_or_else(|| {
+													anyhow::anyhow!(
+														"Block number {} does not exist in the state chain",
+														block_number
+													)
+												})?;
+											Ok(base_rpc_client_c.block_header(hash).await?)
+										}
+									})
+									.map_ok(|header| -> BlockInfo { header.into() }),
+							);
+
+							let start_block = start_from_headers
+								.next()
+								.await
+								.expect("start_from is less than block stream cache number")?;
+
+							Box::pin(
+								start_from_headers
+									.chain(futures::stream::once(futures::future::ready(Ok(
```

Comment:
I think there are some other places we do this, could factor this idea out, into a function on the cached stream that removes the cache and adds it on the front of the stream.

Reply:
The cases are all slightly different it seems, don't think it's super obvious which way to factor it out at this point